aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -87,17 +190,105 @@ public void close() throws IOException {
 
     super.close();
     closed = true;
+    currentStagingFile = null;
 
     try {
       stream.close();
 
+      completeUploads();
+    } finally {
+      stagingFiles.forEach(f -> {
+        if (f.exists() && !f.delete()) {
+          LOG.warn("Could not delete temporary file: {}", f);
+        }
+      });
+    }
+  }
+
+  private void initializeMultiPartUpload() {
+    multipartUploadId = s3.createMultipartUpload(CreateMultipartUploadRequest.builder()
+        .bucket(location.bucket()).key(location.key()).build()).uploadId();
+  }
+
+  private void uploadParts() {
+    // exit if multipart has not been initiated
+    if (multipartUploadId == null) {
+      return;
+    }
+
+    stagingFiles.stream()
+        // do not upload the file currently being written
+        .filter(f -> currentStagingFile == null || !currentStagingFile.equals(f))
+        // do not upload any files that have already been processed
+        .filter(Predicates.not(multiPartMap::containsKey))
+        .forEach(f -> {
+          UploadPartRequest uploadRequest = UploadPartRequest.builder()
+              .bucket(location.bucket())
+              .key(location.key())
+              .uploadId(multipartUploadId)
+              .partNumber(stagingFiles.indexOf(f) + 1)
+              .contentLength(f.length())
+              .build();
+
+          CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
+              () -> {
+                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f));
+                return CompletedPart.builder().eTag(response.eTag()).partNumber(uploadRequest.partNumber()).build();
+              },
+              executorService
+          );
+
+          multiPartMap.put(f, future);
+        });
+  }
+
+  private void completeMultiPartUpload() throws IOException {
+    try {
+      List<CompletedPart> completedParts =
+          multiPartMap.values()
+              .stream()
+              .map(CompletableFuture::join)
+              .sorted(Comparator.comparing(CompletedPart::partNumber))
+              .collect(Collectors.toList());
+
+      s3.completeMultipartUpload(CompleteMultipartUploadRequest.builder()
+          .bucket(location.bucket()).key(location.key())
+          .uploadId(multipartUploadId)
+          .multipartUpload(CompletedMultipartUpload.builder().parts(completedParts).build()).build());
+    } catch (CompletionException e) {
+      abortUpload();
+      throw new IOException("Multipart upload failed for upload id: " + multipartUploadId, e);
+    }
```

Comment:
Here's another place where we can add reliability using `Tasks`. That has a callback for failure that can fail with its own exception, which will be suppressed instead of thrown in place of the `CompletionException`.